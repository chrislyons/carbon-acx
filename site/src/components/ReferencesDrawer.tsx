import { useEffect, useMemo } from 'react';

import { useProfile } from '../state/profile';

type ReferencesDrawerProps = {
  id?: string;
  open: boolean;
  onToggle: () => void;
};

function normaliseReferences(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .map((entry) => (typeof entry === 'string' ? entry.trim() : null))
    .filter((entry): entry is string => Boolean(entry));
}

export function ReferencesDrawer({ id = 'references', open, onToggle }: ReferencesDrawerProps): JSX.Element {
  const { result } = useProfile();

  const references = useMemo(() => normaliseReferences(result?.references), [result]);

  useEffect(() => {
    if (!open) {
      return undefined;
    }
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        event.preventDefault();
        onToggle();
      }
    };
    window.addEventListener('keydown', handleKeyDown);
    return () => window.removeEventListener('keydown', handleKeyDown);
  }, [open, onToggle]);

  return (
    <aside
      id={id}
      aria-labelledby="references-heading"
      aria-live="polite"
      className="rounded-xl border border-slate-800 bg-slate-900/70 p-6 shadow-lg shadow-slate-900/40 backdrop-blur"
    >
      <div className="flex items-start justify-between gap-4">
        <div>
          <h2 id="references-heading" className="text-lg font-semibold text-slate-100">
            References
          </h2>
          <p className="mt-1 text-sm text-slate-400">
            Primary sources supporting the figures. Press <kbd className="rounded bg-slate-800 px-1">Esc</kbd> to close.
          </p>
        </div>
        <button
          type="button"
          aria-expanded={open}
          aria-controls={`${id}-content`}
          onClick={onToggle}
          onKeyDown={(event) => {
            if (event.key.toLowerCase() === 'r') {
              event.preventDefault();
              onToggle();
            }
          }}
          className="inline-flex items-center gap-2 rounded-lg border border-slate-700 px-3 py-2 text-xs font-semibold uppercase tracking-[0.2em] text-slate-200 transition hover:border-slate-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-sky-500"
        >
          <span className="h-1.5 w-1.5 rounded-full bg-sky-400" aria-hidden="true" />
          {open ? 'Collapse' : 'Expand'}
        </button>
      </div>
      <div id={`${id}-content`} hidden={!open} className="mt-6 space-y-4 text-sm text-slate-300">
        {references.length === 0 ? (
          <p className="text-sm text-slate-400">No references available for the current selection.</p>
        ) : (
          <ol className="space-y-3" aria-label="Reference list">
            {references.map((reference, index) => (
              <li
                key={reference}
                className="rounded-lg border border-slate-800/80 bg-slate-950/40 p-3 text-left shadow-inner shadow-slate-900/30"
              >
                <span className="block text-xs uppercase tracking-[0.3em] text-sky-400">[{index + 1}]</span>
                <p className="mt-1 text-sm text-slate-200">{reference.replace(/^\[[0-9]+\]\s*/, '')}</p>
              </li>
            ))}
          </ol>
        )}
      </div>
    </aside>
  );
}
