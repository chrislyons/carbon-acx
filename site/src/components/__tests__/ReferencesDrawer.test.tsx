import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ReferencesDrawer } from '../ReferencesDrawer';

vi.mock('../../state/profile', () => ({
  useProfile: () => ({
    result: {
      references: ['[1] First reference.', '[2] Second reference.']
    }
  })
}));

describe('ReferencesDrawer', () => {
  it('renders references and closes on escape', async () => {
    const user = userEvent.setup();
    const onToggle = vi.fn();
    const { asFragment } = render(<ReferencesDrawer open onToggle={onToggle} />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('First reference.')).toBeInTheDocument();

    await user.keyboard('{Escape}');
    expect(onToggle).toHaveBeenCalled();

    const button = screen.getByRole('button', { name: /collapse/i });
    expect(button.className).toContain('focus-visible');
  });
});
