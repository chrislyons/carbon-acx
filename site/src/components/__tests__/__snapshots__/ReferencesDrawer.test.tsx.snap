// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReferencesDrawer > renders references and closes on escape 1`] = `
<DocumentFragment>
  <aside
    aria-labelledby="references-heading"
    aria-live="polite"
    class="rounded-xl border border-slate-800 bg-slate-900/70 p-6 shadow-lg shadow-slate-900/40 backdrop-blur"
    id="references"
  >
    <div
      class="flex items-start justify-between gap-4"
    >
      <div>
        <h2
          class="text-lg font-semibold text-slate-100"
          id="references-heading"
        >
          References
        </h2>
        <p
          class="mt-1 text-sm text-slate-400"
        >
          Primary sources supporting the figures. Press 
          <kbd
            class="rounded bg-slate-800 px-1"
          >
            Esc
          </kbd>
           to close.
        </p>
      </div>
      <button
        aria-controls="references-content"
        aria-expanded="true"
        class="inline-flex items-center gap-2 rounded-lg border border-slate-700 px-3 py-2 text-xs font-semibold uppercase tracking-[0.2em] text-slate-200 transition hover:border-slate-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-sky-500"
        type="button"
      >
        <span
          aria-hidden="true"
          class="h-1.5 w-1.5 rounded-full bg-sky-400"
        />
        Collapse
      </button>
    </div>
    <div
      class="mt-6 space-y-4 text-sm text-slate-300"
      id="references-content"
    >
      <ol
        aria-label="Reference list"
        class="space-y-3"
      >
        <li
          class="rounded-lg border border-slate-800/80 bg-slate-950/40 p-3 text-left shadow-inner shadow-slate-900/30"
        >
          <span
            class="block text-xs uppercase tracking-[0.3em] text-sky-400"
          >
            [1]
          </span>
          <p
            class="mt-1 text-sm text-slate-200"
          >
            First reference.
          </p>
        </li>
        <li
          class="rounded-lg border border-slate-800/80 bg-slate-950/40 p-3 text-left shadow-inner shadow-slate-900/30"
        >
          <span
            class="block text-xs uppercase tracking-[0.3em] text-sky-400"
          >
            [2]
          </span>
          <p
            class="mt-1 text-sm text-slate-200"
          >
            Second reference.
          </p>
        </li>
      </ol>
    </div>
  </aside>
</DocumentFragment>
`;
